File: java/1123_LowestCommonAncestorOfDeepestLeaves.java
class Solution {
    TreeNode deepestSubtree;
    int deepestDepth = 0;
    public TreeNode lcaDeepestLeaves(TreeNode root) {
        deepestSubtree = root;
        helper(root, 0);
        return deepestSubtree;
    }
    
    public int helper(TreeNode root, int depth) {
        deepestDepth = Math.max(deepestDepth, depth);
        if (root == null) return depth;
        
        int left = helper(root.left, depth + 1), right = helper(root.right, depth + 1);
        if (left == right && left == deepestDepth) deepestSubtree = root;
        return Math.max(left, right);
    }
}

File: cpp/1123_LowestCommonAncestorOfDeepestLeaves.cpp
class Solution {
public:
    int greatest_depth = 0;
    TreeNode* deepest_subtree;
    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        if (!root) return root;
        helper(root, 0);
        return deepest_subtree;
    }
    
    int helper(TreeNode* root, int depth) {
        greatest_depth = max(greatest_depth, depth);
        if (!root) return depth;
        
        int left = helper(root->left, depth + 1), right = helper(root->right, depth + 1);
        
        if (left == right && left == greatest_depth) deepest_subtree = root;
        return max(left, right);
    }
};
